File: core/src/com/sergey/spacegame/common/game/command/Command.java
package com.sergey.spacegame.common.game.command;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import com.sergey.spacegame.client.ui.cursor.CursorOverride;
import com.sergey.spacegame.common.game.Level;
import com.sergey.spacegame.common.lua.LuaUtils;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.Optional;

public final class Command {

	private CommandExecutable executable;
	private boolean requiresInput;
	private boolean requiresTwoInput;
	private String name;
	private String drawableName;
	private String drawableCheckedName;
	private String id;
	private CursorOverride cursor;
	
	public Command(CommandExecutable executable, boolean requiresInput, boolean requiresTwoInput, String name, String drawableName, String drawableCheckedName, CursorOverride cursor) {
		this.executable = executable;
		this.requiresInput = requiresInput;
		this.requiresTwoInput = requiresTwoInput;
		this.name = name;
		this.drawableName = drawableName;
		this.drawableCheckedName = drawableCheckedName;
		this.cursor = cursor;
	}

	/**
	 * @return the executable
	 */
	public CommandExecutable getExecutable() {
		return executable;
	}

	/**
	 * @return the requiresInput
	 */
	public boolean isRequiresInput() {
		return requiresInput;
	}

	/**
	 * @return the requiresTwoInput
	 */
	public boolean isRequiresTwoInput() {
		return requiresTwoInput;
	}
	
	public String getName() {
		return name;
	}
	
	public String getDrawableName() {
		return drawableName;
	}
	
	public String getDrawableCheckedName() {
		return drawableCheckedName;
	}
	
	public String getId() {
		return id;
	}
	
	public CursorOverride getCursor() {
		return cursor;
	}
	
	public void setId(String id) {
		this.id = id;
	}
	
	@Override
	public boolean equals(Object obj) {
		if (!(obj instanceof Command)) return false;
		Command other = (Command)obj;
		return executable.equals(other.executable) && requiresInput == other.requiresInput && requiresTwoInput == other.requiresTwoInput && name.equals(other.name) && drawableName.equals(other.drawableName) && (drawableCheckedName == null ? other.drawableCheckedName == null : drawableCheckedName.equals(other.drawableCheckedName));
	}
	
	@Override
	public int hashCode() {
		return executable.hashCode() << 2 + (requiresInput ? 2 : 0) + (requiresTwoInput ? 1 : 0) + name.hashCode() + drawableName.hashCode() + (drawableCheckedName == null ? 0 : drawableCheckedName.hashCode());
	}
	
	public static class Adapter implements JsonSerializer<Command>, JsonDeserializer<Command> {

		@Override
		public Command deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
			JsonObject obj = json.getAsJsonObject();
			
			String type = obj.getAsJsonPrimitive("type").getAsString();
			CommandExecutable executable;
			if (type.equals("lua")) {

				String lua = obj.get("lua").getAsString();

				try {
					lua = LuaUtils.getLUACode(lua, Level.deserializingFileSystem());
				} catch (IOException e) {
					throw new JsonParseException(e.getMessage(), e);
				}

				executable = new LuaCommandExecutable(lua);
			} else {
				String className = obj.getAsJsonPrimitive("class").getAsString();
				try {
					Class<?> clazz = ClassLoader.getSystemClassLoader().loadClass(className);
					executable = (CommandExecutable) clazz.newInstance();
				} catch (ClassNotFoundException e) {
					throw new JsonParseException("Class " + className + " not found. ", e);
				} catch (InstantiationException | IllegalAccessException e) {
					throw new JsonParseException("Class " + className + " unable to be instantiated. ", e);
				} catch (ClassCastException e) {
					throw new JsonParseException("Class " + className + " does not implement CommandExecutable. ", e);
				}
			}
			
			boolean requiresInput = obj.getAsJsonPrimitive("requiresInput").getAsBoolean();
			boolean requiresTwoInput = obj.getAsJsonPrimitive("requiresTwoInput").getAsBoolean();
			String name = obj.getAsJsonPrimitive("name").getAsString();
			String drawableName = obj.getAsJsonPrimitive("iconName").getAsString();
			String drawableCheckedName = Optional.ofNullable(obj.getAsJsonPrimitive("pressedIconName")).map(JsonPrimitive::getAsString).orElse(null);
			
			CursorOverride cursor = null;
			
			if (obj.has("cursor")) {
				JsonObject cursorObj = obj.getAsJsonObject("cursor");
				String className = cursorObj.get("class").getAsString();
				try {
					Class<?> clazz = ClassLoader.getSystemClassLoader().loadClass(className);
					cursor = context.deserialize(cursorObj, clazz);
				} catch (ClassNotFoundException e) {
					throw new JsonParseException("Class " + className + " not found. ", e);
				}
			}
			
			return new Command(executable, requiresInput, requiresTwoInput, name, drawableName, drawableCheckedName, cursor);
		}

		@Override
		public JsonElement serialize(Command src, Type typeOfSrc, JsonSerializationContext context) {
			JsonObject obj = new JsonObject();
			if (src.getExecutable() instanceof LuaCommandExecutable) {
				obj.addProperty("type", "lua");
				obj.addProperty("lua", ((LuaCommandExecutable)src.getExecutable()).getLua());
			} else {
				obj.addProperty("type", "java");
				obj.addProperty("class", src.getExecutable().getClass().getName());
			}
			obj.addProperty("requiresInput", src.requiresInput);
			obj.addProperty("requiresTwoInput", src.requiresTwoInput);
			obj.addProperty("name", src.name);
			obj.addProperty("iconName", src.drawableName);
			obj.addProperty("pressedIconName", src.drawableCheckedName);
			return obj;
		}

	}

}

File: core/src/com/sergey/spacegame/common/game/command/Command.kt
package com.sergey.spacegame.common.game.command

import com.sergey.spacegame.common.lua.LuaPredicate

/**
 * Represents a base Command
 *
 * @author sergeys
 *
 * @constructor Create a new Command
 *
 * @property executable - the command's executable
 * @property allowMulti - does this command allow multiple targets
 * @property requiresInput - does this command require an input position
 * @property requiresTwoInput - does this command require a secondary input position
 * @property id - the id of this command
 * @property drawableName - the name of the image to use for drawing the command's icon button
 * @property req - the requirements of this command to be enabled
 * @property drawableCheckedName - the name of the image to use for drawing the command's pressed icon button
 * @property orderTag - the tag to use to determine the order count
 */
abstract class Command(val executable: CommandExecutable, val allowMulti: Boolean, val requiresInput: Boolean,
                       val requiresTwoInput: Boolean, val id: String, val drawableName: String,
                       val req: Map<String, LuaPredicate>?,
                       val drawableCheckedName: String?, val orderTag: String?) {
    
    /**
     * The unlocalized name of this command
     */
    val name = "command.$id.name"
    
    /**
     * The unlocalized desciption o this command
     */
    val desc = "command.$id.desc"
    
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Command) return false
        
        if (executable != other.executable) return false
        if (allowMulti != other.allowMulti) return false
        if (requiresInput != other.requiresInput) return false
        if (requiresTwoInput != other.requiresTwoInput) return false
        if (id != other.id) return false
        if (drawableName != other.drawableName) return false
        if (req != other.req) return false
        if (drawableCheckedName != other.drawableCheckedName) return false
        if (orderTag != other.orderTag) return false
        
        return true
    }
    
    override fun hashCode(): Int {
        var result = executable.hashCode()
        result = 31 * result + allowMulti.hashCode()
        result = 31 * result + requiresInput.hashCode()
        result = 31 * result + requiresTwoInput.hashCode()
        result = 31 * result + id.hashCode()
        result = 31 * result + drawableName.hashCode()
        result = 31 * result + (req?.hashCode() ?: 0)
        result = 31 * result + (drawableCheckedName?.hashCode() ?: 0)
        result = 31 * result + (orderTag?.hashCode() ?: 0)
        return result
    }
}

